src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
             return newSize;
         }
 
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up at least all existing entries
+        /// 
+        /// This method can be used to minimize the memory overhead 
+        /// once it is known that no new elements will be added. 
+        /// 
+        /// To completely clear a dictionary and release all memory 
+        /// referenced by the dictionary, execute the following statements:
+        /// 
+        /// dictionary.Clear();
+        /// dictionary.TrimExcess();
+        /// </summary>
+        public void TrimExcess()
+        {
+            TrimExcess(HashHelpers.GetPrime(Count));
+        }
+
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up at least 'capacity' entries without any further expansion of its backing storage
+        /// </summary>
+        public void TrimExcess(int capacity)
+        {
+            if (capacity < Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+
+            if (_entries != null && capacity < _entries.Length)
+            {
+                int[] buckets = new int[capacity];
+                for (int i = 0; i < buckets.Length; i++)
```

Comment:
Can't this just call `Resize`? I do not think we want to have another copy of the rehashing look here.

Reply:
It's a bit different than Resize. I can extract the similarities into a private method which both call though.

please notice that in Resize we are not clearing _freeCount and _freeList

Reply:
That's because of you have bugs in your implementation - you need to round up the capacity to the next prime, etc. Once you fix the bugs, they should be exactly same.

Reply:
`TrimExcess` has to compress the `Dictionary` but `Resize` doesn't.
Calling `Resize` would; currently, also reallocate the arrays. Although when its rehashing its not actually adjusting the arrays so it doesn't need to reallocate - so that could be changed

Reply:
@jkotas I'm addressing getting next prime number as capacity in my next PR.
But the other difference between Resize and TrimExcess I am referring to, is the fact that when we resize, we don't necessarily clear _freeCount, but in TrimExcess we do.

Reply:
> I'm addressing getting next prime number as capacity in my next PR.

I assume that you meant next commit. It does not make sense to delay it to next PR.

Once you start rounding up the capacity to the next prime, you want to be setting `_freeCount` same way as `Resize` because you will have free entries at the end. You want those free entries to be available for future inserts.

Reply:
Alternative would be to go down the freelist and move items from the end into the free spots; then call resize

Reply:
@jkotas re reusing Resize(), it copies all the entry array that has entries (whether free or not) using `            Array.Copy(_entries, 0, entries, 0, count);`. That may be the entire existing Entry array. That would not fit in the smaller array that TrimExcess wants. 

I think it is necessary to do what @benaadams proposes, to defrag all the free entries to the end.

Could you clarify, if I misunderstand?

Reply:
Hmmm.

What would be the impact of replacing the `Array.Copy` with compacting in all cases? It may actually be faster for large dictionaries in some cases because of we would be faulting the array into the cache just once, not twice like with the current code.

The Dictionary code gets stamped many times, especially with AOT. When we are trying to improve Dictionary, we need to spend the extra time to minimize this impact.

Sooner or later, we will need to start the "Power Collections" library where we can go wild with different optimizations, implement collections tailored for different use cases, without worrying about the Dictionary constrains...

Reply:
Compacting would also end up with an Entries array that has potentially better locality. Do you want @maryamariyan to pause this change, and prototype/measure such a change to Resize() ?

Reply:
Yes, I think we should evaluate different options.

Reply:
Thanks. I'll address the other comments in the meantime.